tests/integration/targets/ec2_vpc_nat_gateway/tasks/main.yml

```diff
@@ -3,80 +3,527 @@
+  #       tag_one: '{{ resource_prefix }} One'
+  #     wait: yes
+  #   register: update_tags_ngw
+  #   check_mode: yes
+
+  # - name: Assert tags would be added - CHECK_MODE
+  #   assert:
+  #     that:
+  #         - update_tags_ngw.changed
+
+
+  # ============================================================
+  - name: Remove all tags
+    ec2_vpc_nat_gateway:
+      subnet_id: "{{ subnet_id }}"
+      allocation_id: "{{ allocation_id }}"
+      tags: {}
+    register: delete_tags_ngw
+
+  - name: assert tags would be removed
+    assert:
+      that:
+        - delete_tags_ngw.changed
+  
+  # - name: Remove all tags - CHECK_MODE
+  #   ec2_vpc_nat_gateway:
+  #     subnet_id: "{{ subnet_id }}"
+  #     allocation_id: "{{ allocation_id }}"
+  #     tags: {}
+  #   register: delete_tags_ngw
+  #   check_mode: yes
+
+  # - name: assert tags would be removed - CHECK_MODE
+  #   assert:
+  #     that:
+  #         - delete_tags_ngw.changed
+
+  
+  
+  # ============================================================
+  - name: Update with CamelCase tags
+    ec2_vpc_nat_gateway:
+      if_exist_do_not_create: yes
+      subnet_id: "{{ subnet_id }}"
+      allocation_id: "{{ allocation_id }}"
+      purge_tags: no
+      tags:
+        "lowercase spaced": 'hello cruel world ❤️'
+        "Title Case": 'Hello Cruel World ❤️'
+        CamelCase: 'SimpleCamelCase ❤️'
+        snake_case: 'simple_snake_case ❤️'
+      wait: yes
+    register: update_tags_ngw
+
+  - name: Assert tags would be added
+    assert:
+      that:
+         - update_tags_ngw.changed
+         - update_tags_ngw.tags["lowercase spaced"] == 'hello cruel world ❤️'
+         - update_tags_ngw.tags["Title Case"] == 'Hello Cruel World ❤️'
+         - update_tags_ngw.tags["CamelCase"] == 'SimpleCamelCase ❤️'
+         - update_tags_ngw.tags["snake_case"] == 'simple_snake_case ❤️'
+
+
+  # - name: Update with CamelCase tags - CHECK_MODE
+  #  ec2_vpc_nat_gateway:
+  #    if_exist_do_not_create: yes
+  #    subnet_id: "{{ subnet_id }}"
+  #    allocation_id: "{{ allocation_id }}"
+  #    purge_tags: no
+  #    tags:
+  #      "lowercase spaced": 'hello cruel world ❤️'
+  #      "Title Case": 'Hello Cruel World ❤️'
+  #      CamelCase: 'SimpleCamelCase ❤️'
+  #      snake_case: 'simple_snake_case ❤️'
+  #    wait: yes
+  # register: update_tags_ngw
+
+  #- name: Assert tags would be added - CHECK_MODE
+  #  assert:
+  #    that:
+  #       - update_tags_ngw.changed
+  #       - update_tags_ngw.tags["lowercase spaced"] == 'hello cruel world ❤️'
+  #       - update_tags_ngw.tags["Title Case"] == 'Hello Cruel World ❤️'
+  #       - update_tags_ngw.tags["CamelCase"] == 'SimpleCamelCase ❤️'
+  #       - update_tags_ngw.tags["snake_case"] == 'simple_snake_case ❤️'
+
+  # ============================================================
+  
+  
+  always:
+  - name: Tidy up NAT gateway
+    ec2_vpc_nat_gateway:
+      subnet_id: "{{ subnet_id }}"
+      nat_gateway_id: "{{ nat_gateway_id }}"
```

Comment:
You're using `'{{ nat_gateway_id }}'` here, but that var refers to the NAT GW that was deleted in the task at lines 255-260. The active NAT GW at the end of the tests is the one created at lines 281-289 - the ID never gets assigned to a new var, so it still exists when you try to delete the other resources.

You'll want to loop through and delete both NAT Gateway IDs just to be safe (if something in the test fails before line 255, the `always` block should still try to clean the first one up).  You can either store both IDs in their own vars, or put a `ec2_vpc_nat_gateway_info` task with a `vpc-id` filter into the `always` block and use the results of that

Reply:
@jillr Thank you Jill. I fixed this by making sure all dependencies were released before removing the VPC.